### ISSUE TITLE
uuid1 is not a thread-safe way of generating session_id and can result in SESSION_IS_LOCKED

### ISSUE THREAD

**Issue by guillaumematheron** [ClickHouse/clickhouse-connect#194]
The code at fault here is https://github.com/ClickHouse/clickhouse-connect/blob/main/clickhouse_connect/driver/httpclient.py#L126

The python documentation states that uuid1 :

> Generate a UUID from a host ID, sequence number, and the current time. If node is not given, getnode() is used to obtain the hardware address. If clock_seq is given, it is used as the sequence number; otherwise a random 14-bit sequence number is chosen.

However this does not seem to be accurate. On my platform (Ubuntu 22.04, python3.8) I get duplicates when using multiprocessing.

Even if it was accurate, uuid1 does not guarantee that it is thread_safe:

> Depending on support from the underlying platform, [uuid1()](https://docs.python.org/3/library/uuid.html#uuid.uuid1) may or may not return a “safe” [UUID](https://docs.python.org/3/library/uuid.html#uuid.UUID). A safe UUID is one which is generated using synchronization methods that ensure no two processes can obtain the same UUID. All instances of UUID have an is_safe attribute which relays any information about the UUID’s safety, using this enumeration:

Note that on my platform the flag is_safe is set even when the UUIDs are identical in several processes -> therefore this is not a reliable flag. Even if it was, it's unclear what the strategy would be when we get an unsafe UUID.

A workaround is to supply your own uuid4 to get_client.

In conclusion, I think we should replace uuid1 with uuid4 in this context.

**Comment by genzgd** [comment]
I actually faced exactly this issue yesterday when working on some test code.  The problem is not uuid1(), which is fine -- the 14 bits of randomness plus the time sequence always generated a unique id but I get kept getting SESSION LOCKED errors.  The problem was in error handling with streaming queries that manifested as a SESSION LOCKED error because the streaming query was not correctly closed.

Do you have a small reproducible example?  Can you validate that uuid4() fixes the issue?

**Comment by genzgd** [comment]
Those are not sequential.   Without digging too far into the various bits in a uuid1, the only part that's sequential is the sequence number.  If you can't create identical uuids even when calling that uuid1 quickly, that's a good sign it's not the problem.  I'm pretty confident that new clients get a new, unique session_id:

```python
#!/usr/bin/env python -u

import time
from multiprocessing import Process

import clickhouse_connect


def new_client():
    client = clickhouse_connect.get_client()
    print(client.params['session_id'])
    time.sleep(5000)
    print('exiting')


def start_stuff():
    for x in range(4):
        Process(target=new_client).start()


if __name__ == '__main__':
    start_stuff()


e0d95f50-0168-11ee-80c9-62b66330619f
e0e112c2-0168-11ee-890b-62b66330619f
e0e2dc56-0168-11ee-bf93-62b66330619f
e0e4677e-0168-11ee-ba5d-62b66330619f
```

I think you are much closer by noting that you need a new pool manager per client.  The means your clients will never be sharing a TCP connection, and I think the core of the problem is somehow related to the connection not being closed correctly (and therefore ClickHouse is not considering the query "complete") under certain conditions -- in particular I see this with streaming inserts.

Just to be clear, I'm not disputing there's a bug.  I just think it's not as simple as a shared session_id.

**Comment by guillaumematheron** [comment]
TLDR : I managed to get a MWE involving only python at the end of this comment.


Ok, since I can't get a MWE yet, I dug into the main code.

```python
        self._client = clickhouse_connect.get_client(
            host=hostname,
            username=clickhouse_user,
            password=clickhouse_pass,
            port=port,
            pool_mgr=PoolManager(),  # Otherwise we get horrible bugs :'(
            # settings={"session_id": str(uuid.uuid4())},  # See https://github.com/ClickHouse/clickhouse-connect/issues/194
        )

```

Then I get plenty of those:

```
ERROR:clickhouse_connect.driver.httpclient:Code: 373. DB::Exception: Session 6877db77-0182-11ee-ae75-9fccb38c555d is locked by a concurrent client. (SESSION_IS_LOCKED) (version 23.4.2.11 (official build))

ERROR:clickhouse_connect.driver.httpclient:Code: 373. DB::Exception: Session 6877db77-0182-11ee-ae75-9fccb38c555d is locked by a concurrent client. (SESSION_IS_LOCKED) (version 23.4.2.11 (official build))

ERROR:clickhouse_connect.driver.httpclient:Code: 373. DB::Exception: Session 6877db77-0182-11ee-ae75-9fccb38c555d is locked by a concurrent client. (SESSION_IS_LOCKED) (version 23.4.2.11 (official build))
```

So I'll just go into the HTTPClient ctor and replace:

```
        if session_id:
            ch_settings['session_id'] = session_id
        elif 'session_id' not in ch_settings and common.get_setting('autogenerate_session_id'):
            ch_settings['session_id'] = str(uuid.uuid1())
```

with

```
        if session_id:
            ch_settings['session_id'] = session_id
        elif 'session_id' not in ch_settings and common.get_setting('autogenerate_session_id'):
            sid = uuid.uuid1()
            import time
            import os
            print(f"sid={str(sid)}, is_safe={sid.is_safe} time={time.time()}, pid={os.getpid()}")
            ch_settings['session_id'] = str(sid)
```

Now we get:

```
sid=9baa869a-0189-11ee-ae75-9fccb38c555d, is_safe=SafeUUID.safe time=1685740102.8856783, pid=448570
sid=9ba05919-0189-11ee-ae75-9fccb38c555d, is_safe=SafeUUID.safe time=1685740102.954511, pid=448614
sid=9ba05919-0189-11ee-ae75-9fccb38c555d, is_safe=SafeUUID.safe time=1685740102.954622, pid=448615
sid=9ba05919-0189-11ee-ae75-9fccb38c555d, is_safe=SafeUUID.safe time=1685740102.9547586, pid=448616
sid=9ba05919-0189-11ee-ae75-9fccb38c555d, is_safe=SafeUUID.safe time=1685740102.9547575, pid=448617
sid=9ba05919-0189-11ee-ae75-9fccb38c555d, is_safe=SafeUUID.safe time=1685740102.954903, pid=448618
sid=9ba05919-0189-11ee-ae75-9fccb38c555d, is_safe=SafeUUID.safe time=1685740102.9549818, pid=448619
sid=9ba05919-0189-11ee-ae75-9fccb38c555d, is_safe=SafeUUID.safe time=1685740102.955067, pid=448620
sid=9ba05919-0189-11ee-ae75-9fccb38c555d, is_safe=SafeUUID.safe time=1685740102.9551823, pid=448621
sid=9ba05919-0189-11ee-ae75-9fccb38c555d, is_safe=SafeUUID.safe time=1685740102.957512, pid=448623
sid=9ba05919-0189-11ee-ae75-9fccb38c555d, is_safe=SafeUUID.safe time=1685740102.9575124, pid=448622
```

Ok let's try to know more about this:

```python
        if session_id:
            ch_settings['session_id'] = session_id
        elif 'session_id' not in ch_settings and common.get_setting('autogenerate_session_id'):
            sid = uuid.uuid1()
            sid2 = uuid.uuid1()
            import time
            import os
            print(f"sid ={str(sid)}, is_safe={sid.is_safe} time={time.time()}, pid={os.getpid()}")
            print(f"sid2={str(sid2)}, is_safe={sid2.is_safe} time={time.time()}, pid={os.getpid()}")
            print()
            ch_settings['session_id'] = str(sid)
```

```
sid =13d86a59-018a-11ee-ae75-9fccb38c555d, is_safe=SafeUUID.safe time=1685740304.655585, pid=449272
sid2=13d86a5a-018a-11ee-ae75-9fccb38c555d, is_safe=SafeUUID.safe time=1685740304.6556408, pid=449272

sid =13d86a59-018a-11ee-ae75-9fccb38c555d, is_safe=SafeUUID.safe time=1685740304.655759, pid=449273
sid2=13d86a5a-018a-11ee-ae75-9fccb38c555d, is_safe=SafeUUID.safe time=1685740304.6558123, pid=449273
```

So within a single process, subsequent uuid calls are sequential, but the time component is off.


Let's go back to trying a MWE:

```python
def worker(_):
    print(uuid.uuid1())

with multiprocessing.Pool(4) as p:
    r = p.imap_unordered(worker, list(range(4)))
    for x in r:
        _ = x
```

```
da4fd11a-018a-11ee-ae7b-9fccb38c555d                                                                                                                                                                                 
da4fd39a-018a-11ee-ae7c-9fccb38c555d                                                                                                                                                                                 
da4fd642-018a-11ee-ae7d-9fccb38c555d                                                                                                                                                                                 
da4fd8e0-018a-11ee-ae7e-9fccb38c555d
````

This is working as expected (although again, the python doc says at list some component should be random whereas this is pretty clearly a timer and a sequence).

Since in my main program I also use a client for the main process, let's add a call to uuid1 early:

```python3
import uuid
import multiprocessing

print(uuid.uuid1())
print(uuid.uuid1())

print()

def worker(_):
    print(uuid.uuid1())

with multiprocessing.Pool(4) as p:
    r = p.imap_unordered(worker, list(range(4)))
    for x in r:
        _ = x
```

```
4424e35a-018b-11ee-ae7e-9fccb38c555d                                                                                                                                                                                 
4424e35b-018b-11ee-ae7e-9fccb38c555d                                                                                                                                                                                 
                                                                                                                                                                                                                     
4424e35c-018b-11ee-ae7e-9fccb38c555d                                                                                                                                                                                 
4424e35c-018b-11ee-ae7e-9fccb38c555d                                                                                                                                                                                 
4424e35c-018b-11ee-ae7e-9fccb38c555d                                                                                                                                                                                 
4424e35c-018b-11ee-ae7e-9fccb38c555d     
```

Here it is ! It's a state thing within uuid, I'll try to debug it further and report back (I'm using Python 3.8.13)

**Comment by genzgd** [comment]
Okay, I'm convinced.  uuid1 does not seem to be thread safe in your environment in multiprocessing mode.  Are you using Linux?  Do you happen to set the start_method mode for multiprocessing?   It's `fork` by default which has some known issues.  https://bugs.python.org/issue40379.  Does it work with `set_start_mode('spawn')`

One more question -- does uuid4 fix it? 

I'm thinking that the PoolManager may also not be thread safe in a multiprocessing environment, if you also needed that to avoid weird bugs.  I suspect you do need a separate PoolManager per client.  I will see if there's some way to do that automatically; in the meantime instead of just `PoolManager()` I would recommend calling `httputil.get_pool_manager()` which sets some important socket options and defaults.

**Comment by genzgd** [comment]
Btw thanks for your persistence in digging into and reproducing the problem.

**Comment by genzgd** [comment]
I didn't notice the fork/spawn issue had been moved to Github.  More discussion and context here:  https://github.com/python/cpython/issues/84559

**Comment by guillaumematheron** [comment]
Yep, definitely the problem is with spawn/fork:

By switching to this MWE, I get different UUIDs:

```python
import uuid
import multiprocessing

def worker(_):
    print(uuid.uuid1())

if __name__ == "__main__":
    multiprocessing.set_start_method('spawn')
    print(uuid.uuid1())
    print(uuid.uuid1())

    print()

    with multiprocessing.Pool(4) as p:
        r = p.imap_unordered(worker, list(range(4)))
        for x in r:
            _ = x
```

```
d6dd92ac-0238-11ee-a9b2-134f3c9f0156
d6dd92ad-0238-11ee-a9b2-134f3c9f0156

d6e70274-0238-11ee-a9b2-134f3c9f0156
d6e70275-0238-11ee-a9b2-134f3c9f0156
d6e70276-0238-11ee-a9b2-134f3c9f0156
d6e70277-0238-11ee-a9b2-134f3c9f0156
```

I still don't like the fact that the documentation says there should be 14 bits of randomness, but that's a documentation issue only.

Switching to uuid4 also works as expected.

So I suppose the first call to uuid1 sets up a global state that gets copied to each process, which then loses its uniqueness. Combined with the use of the c function uuid_generate_time_safe, which apparently uses a timestamp with a resolution that is not high enough, this causes duplicate uuids.

So the issue is mostly with python, I'm not sure how you want to proceed with clickhouse-connect. For me it makes a lot of sense to switch to uuid4, since I don't expect any changes in defaults for multiprocessing to be backported to earlier python versions ?

**Comment by genzgd** [comment]
I suspect there are indeed 14 bits of randomness as coded, but maybe each process ends up  with the same random (even with spawn vs fork) meaning you just get the sequence difference.  

In any case I'm fine with switching to uuid4 if it reliably generates a truly random/unique UUID regardless of the multiprocessing environment.  The risk of collision remains basically non-existent, while there are evidently fewer opportunities for weird multiprocess bugs.  I've already added the fixes in the branch for the 0.6.0 release.

**Comment by guillaumematheron** [comment]
Since two consecutive calls to uuid1 give very similar results (with some bytes constant and some bytes incremented), I don't think there is any random component. I guess the initialization could be random, but consecutive runs also give very similar results :


```
$ python main.py 
f3a2b3f8-036a-11ee-a9b2-134f3c9f0156
f3a2b3f9-036a-11ee-a9b2-134f3c9f0156

f3a2b3fa-036a-11ee-a9b2-134f3c9f0156
f3a2b3fa-036a-11ee-a9b2-134f3c9f0156
f3a2b3fb-036a-11ee-a9b2-134f3c9f0156
f3a2b3fa-036a-11ee-a9b2-134f3c9f0156

$ python main.py 
f3cda69e-036a-11ee-a9b2-134f3c9f0156
f3cda69f-036a-11ee-a9b2-134f3c9f0156

f3cda6a0-036a-11ee-a9b2-134f3c9f0156
f3cda6a0-036a-11ee-a9b2-134f3c9f0156
f3cda6a0-036a-11ee-a9b2-134f3c9f0156
f3cda6a0-036a-11ee-a9b2-134f3c9f0156
```

I confirmed this by calling uuid1 in a single thread 1000 times in a loop, and strace reports only two calls to getrandom (8 bytes, then 24).

The 8 bytes then 24 bytes happen even for an empty program, and then 32 extra bytes seem to be triggered by `import multiprocessing`.

FWIW, the underlying function called seems to be [uuid_generate_time_safe](https://github.com/util-linux/util-linux/blob/master/libuuid/src/gen_uuid.c), and the man is pretty obscure on how this function works, but it looks like random values are only used if a network address is not available.

**Comment by pitrou** [comment]
IMO you're overthinking this. Just use [`random.SystemRandom`](https://docs.python.org/3/library/random.html#random.SystemRandom) or the [secrets modules](https://docs.python.org/3/library/secrets.html).

**Comment by genzgd** [comment]
@guillaumematheron I get somewhat different uuid1 results on my mac, but I was under the impression that the 14 bit "clock sequence" was random, but it is not (nor is it documented that way).  In any case, although uuid is the instinctive way to get -- well, a UUID -- at this point I'm inclined to use `secrets.SecureRandom` as suggested just so this doesn't come up again in some other weird random number generating context.
